File: Python/OCP_demo.py
from abc import ABC, abstractmethod


class IDiscount(ABC):
    @abstractmethod
    def get_discounted_price(self, amount: int) -> float:
        pass


class Festival(IDiscount):
    discount = 0.25

    def get_discounted_price(self, amount: int) -> float:
        return amount * self.discount


class Holi(IDiscount):
    discount = 0.25

    def get_discounted_price(self, amount: int) -> float:
        return amount * self.discount


class PaymentService:
    def __init__(self, discount: IDiscount):
        self.discount = discount

    def pay(self, amount: int) -> None:
        final = self.discount.get_discounted_price(amount)
        print(final)


if __name__ == "__main__":
    service = PaymentService(Festival())  # solution is Factory Method design Pattern

File: Golang/OCP_demo.go
package main

import "fmt"

type IDiscount interface {
	GetDiscountedPrice(amount int) float32
}

type Festival struct {
	discount float32
}

func NewFestival() *Festival {
	return &Festival{discount: 0.25}
}

func (f *Festival) GetDiscountedPrice(amount int) float32 {
	return float32(amount) * f.discount
}

type Holi struct {
	discount float32
}

func NewHoli() *Holi {
	return &Holi{discount: 0.25}
}

func (h *Holi) GetDiscountedPrice(amount int) float32 {
	return float32(amount) * h.discount
}

type PaymentService struct {
	discount IDiscount
}

func NewPaymentService(discount IDiscount) *PaymentService {
	return &PaymentService{discount: discount}
}

func (p *PaymentService) Pay(amount int) {
	final := p.discount.GetDiscountedPrice(amount)
	fmt.Println(final)
}

func main() {
	service := NewPaymentService(NewFestival()) // solution is Factory Method design Pattern
	_ = service
}
